Avoid calling scanTagName() in HTMLFastPathParser::parseContainerElement()

cdumez opened the pull request:
#### c727638623487a921b8bee06568220f69965db10
<pre>
Avoid calling scanTagName() in HTMLFastPathParser::parseContainerElement()
<a href="https://bugs.webkit.org/show_bug.cgi?id=254920">https://bugs.webkit.org/show_bug.cgi?id=254920</a>

Reviewed by Ryosuke Niwa.

parseContainerElement() used to call scanTagName(), which would read all the
characters of the tag and convert them to lowercase. Then it would convert
the Span of characters to an ElementName enum.

However, in this case, we&apos;re parsing the end tag and we know it has to match
the start tag. If it doesn&apos;t then the element wouldn&apos;t be valid. As a result,
we can just call skipCharactersExactly() in the fast path, which will do a
simple memcmp() and succeed most of the time. In the slow case, we fall back
to calling skipLettersExactlyIgnoringASCIICase() in case the end tag is not
lowercase.

This is slightly more efficient.

* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::HTMLFastPathParser::parseContainerElement):
* Source/WebCore/html/parser/ParsingUtilities.h:
(WebCore::skipLettersExactlyIgnoringASCIICase):
(WebCore::skipCharactersExactly): Deleted.

Canonical link: <a href="https://commits.webkit.org/262520@main">https://commits.webkit.org/262520@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80a4d3d1c8150528248e6ef03cdd20bbca1aea05

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1768 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1799 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1857 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2697 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1893 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1750 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1870 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1866 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1658 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1786 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1604 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1586 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2535 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1597 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1572 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1499 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1628 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1615 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2698 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1658 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1476 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1581 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1579 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/434 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1720 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -340,18 +343,22 @@ class HTMLFastPathParser {
 
         struct Br : VoidTag<HTMLBRElement, PermittedParents::PhrasingOrFlowContent> {
             static constexpr ElementName tagName = ElementName::HTML_br;
+            static constexpr Char tagNameCharacters[] = { 'b', 'r' };
```
Can we use "br" instead? Ditto for other multi-character const expressions.

cdumez replied:
Char can be a LChar or a UChar (depending on whether the buffer being parsed is 8bit or not).

rniwa replied:
Oh, I see. That's unfortunate.

cdumez replied:
I considered using ASCIILiteral but then the code would be slower when dealing with a 16bit input since it would have to iterate over the string manually instead of using memcmp().

cdumez committed:
Avoid calling scanTagName() in HTMLFastPathParser::parseContainerElement()
https://bugs.webkit.org/show_bug.cgi?id=254920

Reviewed by Ryosuke Niwa.

parseContainerElement() used to call scanTagName(), which would read all the
characters of the tag and convert them to lowercase. Then it would convert
the Span of characters to an ElementName enum.

However, in this case, we're parsing the end tag and we know it has to match
the start tag. If it doesn't then the element wouldn't be valid. As a result,
we can just call skipCharactersExactly() in the fast path, which will do a
simple memcmp() and succeed most of the time. In the slow case, we fall back
to calling skipLettersExactlyIgnoringASCIICase() in case the end tag is not
lowercase.

This is slightly more efficient.

* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::HTMLFastPathParser::parseContainerElement):
* Source/WebCore/html/parser/ParsingUtilities.h:
(WebCore::skipLettersExactlyIgnoringASCIICase):
(WebCore::skipCharactersExactly): Deleted.

Canonical link: https://commits.webkit.org/262520@main
